### ISSUE TITLE
plugin: needs at least one argument

### ISSUE THREAD

**Issue by sfuerte** [hashicorp/packer#11234]
#### Overview of the Issue
Upgrading an open-source plugin to use the latest SDK and API. The code is at https://github.com/sfuerte/packer-builder-qemu-chroot/tree/packer_1_7

While it does compile fine, but when trying to run, it fails with the subject error.

#### Reproduction Steps

- clone the above-mentioned repo
- build a plugin
- try to use the `examples/ubuntu.json` (even without downloading any extra images)

### Packer version

From `packer version`, as well as the plugin info:
```sh
$ /opt/packer_1.7.4/packer version
Packer v1.7.4

$ ./packer-builder-qemu-chroot describe
{"version":"1.1.0-u1-dev","sdk_version":"0.2.3","api_version":"x5.0","builders":["qemu-chroot"],"post_processors":[],"provisioners":[],"datasources":[]}
```

### Simplified Packer Buildfile

```json
{
  "builders": [{
      "name": "build-test",
      "type": "qemu-chroot",
      "source_image": "ubuntu-16.04-server-cloudimg-amd64-disk1.img",
      "image_name": "ubuntu-16.04.img",
      "compression": true
    }],
  "provisioners": [    {
      "type": "shell",
      "inline": ["apt update"]
    }]
}
```

### Operating system and Environment details

```sh
$ lsb_release -d
Description:	Debian GNU/Linux 10 (buster)
```

### Log Fragments and crash.log files

the `stdout`:
```sh
$ sudo PACKER_LOG_PATH="/home/debian/packer-builder-qemu-chroot/packerlog.txt" PACKER_LOG=1  /opt/packer_1.7.4/packer build -debug examples/ubuntu.json
Error: Failed to initialize build "build-test"

error initializing builder 'qemu-chroot': plugin exited before we could connect


Debug mode enabled. Builds will not be parallelized.

==> Wait completed after 7 microseconds

==> Builds finished but no artifacts were created.
```

`packerlog.txt`:
```
2021/08/30 21:16:52 [INFO] Packer version: 1.7.4 [go1.16.6 linux amd64]
2021/08/30 21:16:52 [TRACE] discovering plugins in /opt/packer_1.7.4
2021/08/30 21:16:52 [DEBUG] Discovered plugin: qemu-chroot = /opt/packer_1.7.4/packer-builder-qemu-chroot
2021/08/30 21:16:52 [INFO] using external builders: [qemu-chroot]
2021/08/30 21:16:52 [TRACE] discovering plugins in /root/.packer.d/plugins
2021/08/30 21:16:52 [TRACE] discovering plugins in .
2021/08/30 21:16:52 [DEBUG] Discovered plugin: qemu-chroot = /home/debian/packer-builder-qemu-chroot/packer-builder-qemu-chroot
2021/08/30 21:16:52 [INFO] using external builders: [qemu-chroot]
2021/08/30 21:16:52 [INFO] PACKER_CONFIG env var not set; checking the default config file path
2021/08/30 21:16:52 [INFO] PACKER_CONFIG env var set; attempting to open config file: /root/.packerconfig
2021/08/30 21:16:52 [WARN] Config file doesn't exist: /root/.packerconfig
2021/08/30 21:16:52 [INFO] Setting cache directory: /home/debian/packer-builder-qemu-chroot/packer_cache
2021/08/30 21:16:52 [TRACE] Starting external plugin /home/debian/packer-builder-qemu-chroot/packer-builder-qemu-chroot
2021/08/30 21:16:52 Starting plugin: /home/debian/packer-builder-qemu-chroot/packer-builder-qemu-chroot []string{"/home/debian/packer-builder-qemu-chroot/packer-builder-qemu-chroot"}
2021/08/30 21:16:52 Waiting for RPC address for: /home/debian/packer-builder-qemu-chroot/packer-builder-qemu-chroot
2021/08/30 21:16:52 packer-builder-qemu-chroot plugin: needs at least one argument
2021/08/30 21:16:52 /home/debian/packer-builder-qemu-chroot/packer-builder-qemu-chroot: plugin process exited
2021/08/30 21:16:52 ui error: Error: Failed to initialize build "build-test"

error initializing builder 'qemu-chroot': plugin exited before we could connect

2021/08/30 21:16:52 ui: Debug mode enabled. Builds will not be parallelized.
2021/08/30 21:16:52 Build debug mode: true
2021/08/30 21:16:52 Force build: false
2021/08/30 21:16:52 On error:
2021/08/30 21:16:52 Waiting on builds to complete...
2021/08/30 21:16:52 ui:
==> Wait completed after 7 microseconds
2021/08/30 21:16:52 ui:
==> Builds finished but no artifacts were created.
2021/08/30 21:16:52 [INFO] (telemetry) Finalizing.
2021/08/30 21:16:53 waiting for all plugin processes to complete...
```

**Comment by nywilken** [comment]
Hi @sfuerte thanks for reaching out. It looks like you are using the previous naming convention for Packer plugins which may be the issue here. By default any plugin using the naming convention `packer-[builder|provisioner|post-processor]-*` will get loaded automatically by Packer and get started as if it was a single component plugin. Which can be seen in the log lines: 

```
2021/08/30 21:16:52 [TRACE] Starting external plugin /home/debian/packer-builder-qemu-chroot/packer-builder-qemu-chroot
2021/08/30 21:16:52 Starting plugin: /home/debian/packer-builder-qemu-chroot/packer-builder-qemu-chroot []string{"/home/debian/packer-builder-qemu-chroot/packer-builder-qemu-chroot"}
2021/08/30 21:16:52 Waiting for RPC address for: /home/debian/packer-builder-qemu-chroot/packer-builder-qemu-chroot
```

When using the new multi-component plugin type like you are doing for your plugin the expected naming convention is `packer-plugin-*`. 

In your case it could be `packer-plugin-qemu-chroot`, but it would require that you use `plugin.DEFAULT_NAME` when registering the plugin as in `pps.RegisterBuilder(plugin.DEFAULT_NAME, new(chroot.Builder)))`

If you try building and running your plugin as `packer-plugin-qemu-chroot` do you still run into the same issue?

**Comment by sfuerte** [comment]
Hi @nywilken, 

Thanks a lot for the prompt reply. 
Interestingly, I was following the original source - Amazon's `chroot` plugin, https://github.com/hashicorp/packer-plugin-amazon/blob/main/main.go - which doesn't use `plugin.DEFAULT_NAME` but a defined string. 
Though their module/folder naming is with `packer-plugin-` instead of `packer-builder-`. 


Anyway, I'm passed that stage and troubleshooting the next one but that might be SDK incompatibilities:

```
panic: runtime error: invalid memory address or nil pointer dereference
2021/09/02 02:21:49 packer-plugin-qemu-chroot plugin: [signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x61cea5]
2021/09/02 02:21:49 packer-plugin-qemu-chroot plugin:
2021/09/02 02:21:49 packer-plugin-qemu-chroot plugin: goroutine 37 [running]:
2021/09/02 02:21:49 packer-plugin-qemu-chroot plugin: github.com/hashicorp/packer-plugin-sdk/packerbuilderdata.(*GeneratedData).Put(0x0, {0xd9d2fc, 0x6}, {0xc5fa20, 0xc000305820})
2021/09/02 02:21:49 packer-plugin-qemu-chroot plugin: 	/home/debian/go/pkg/mod/github.com/hashicorp/packer-plugin-sdk@v0.2.3/packerbuilderdata/generated_data.go:26 +0x45
2021/09/02 02:21:49 packer-plugin-qemu-chroot plugin: github.com/summerwind/packer-plugin-qemu-chroot/qemu/chroot.(*StepPrepareDevice).Run(0xc000010098, {0xda140c, 0x40b975}, {0xf707e8, 0xc0005aa360})
2021/09/02 02:21:49 packer-plugin-qemu-chroot plugin: 	/home/debian/packer-plugin-qemu-chroot/qemu/chroot/step_prepare_device.go:51 +0x238
2021/09/02 02:21:49 packer-plugin-qemu-chroot plugin: github.com/hashicorp/packer-plugin-sdk/multistep.(*BasicRunner).Run(0xc0005aa450, {0xf70388, 0xc000032140}, {0xf707e8, 0xc0005aa360})
2021/09/02 02:21:49 packer-plugin-qemu-chroot plugin: 	/home/debian/go/pkg/mod/github.com/hashicorp/packer-plugin-sdk@v0.2.3/multistep/basic_runner.go:70 +0x2fe
2021/09/02 02:21:49 packer-plugin-qemu-chroot plugin: github.com/hashicorp/packer-plugin-sdk/multistep.(*DebugRunner).Run(0xc0005aa420, {0xf70388, 0xc000032140}, {0xf707e8, 0xc0005aa360})
2021/09/02 02:21:49 packer-plugin-qemu-chroot plugin: 	/home/debian/go/pkg/mod/github.com/hashicorp/packer-plugin-sdk@v0.2.3/multistep/debug_runner.go:84 +0x1b8
2021/09/02 02:21:49 packer-plugin-qemu-chroot plugin: github.com/summerwind/packer-plugin-qemu-chroot/qemu/chroot.(*Builder).Run(0xc00052c540, {0xf70388, 0xc000032140}, {0xf76798, 0xc0005aa2a0}, {0xf5a880, 0xc000044040})
2021/09/02 02:21:49 packer-plugin-qemu-chroot plugin: 	/home/debian/packer-plugin-qemu-chroot/qemu/chroot/builder.go:173 +0x4d9
2021/09/02 02:21:49 packer-plugin-qemu-chroot plugin: github.com/hashicorp/packer-plugin-sdk/rpc.(*BuilderServer).Run(0xc00030b8c0, 0x30c0d0, 0xc00030c0b0)
2021/09/02 02:21:49 packer-plugin-qemu-chroot plugin: 	/home/debian/go/pkg/mod/github.com/hashicorp/packer-plugin-sdk@v0.2.3/rpc/builder.go:117 +0x1e3
2021/09/02 02:21:49 packer-plugin-qemu-chroot plugin: reflect.Value.call({0xc000130300, 0xc000010080, 0x13}, {0xd9b9a8, 0x4}, {0xc000593ef8, 0x3, 0x3})
2021/09/02 02:21:49 packer-plugin-qemu-chroot plugin: 	/usr/local/go/src/reflect/value.go:543 +0x814
2021/09/02 02:21:49 packer-plugin-qemu-chroot plugin: reflect.Value.Call({0xc000130300, 0xc000010080, 0xc000580000}, {0xc000554ef8, 0x3, 0x3})
2021/09/02 02:21:49 packer-plugin-qemu-chroot plugin: 	/usr/local/go/src/reflect/value.go:339 +0xc5
2021/09/02 02:21:49 packer-plugin-qemu-chroot plugin: net/rpc.(*service).call(0xc00030b900, 0xc00013a0a9, 0x0, 0xc0000360f0, 0xc00019e200, 0x5f0985, {0xc5fc20, 0xc00030c00c, 0x5f0906}, {0xc488e0, ...}, ...)
2021/09/02 02:21:49 packer-plugin-qemu-chroot plugin: 	/usr/local/go/src/net/rpc/server.go:377 +0x239
2021/09/02 02:21:49 packer-plugin-qemu-chroot plugin: created by net/rpc.(*Server).ServeCodec
2021/09/02 02:21:49 packer-plugin-qemu-chroot plugin: 	/usr/local/go/src/net/rpc/server.go:474 +0x405
```

also, another weirdness is a pause after each step:
```
==> build-test: Pausing after run of step 'StepPrepareOutputDir'. Press enter to continue.
==> build-test: Copying source image...
==> build-test: Pausing after run of step 'StepPrepareImage'. Press enter to continue.
==> build-test: Finding available device...
==> build-test: Pausing before cleanup of step 'StepPrepareImage'. Press enter to continue.
==> build-test: Pausing before cleanup of step 'StepPrepareOutputDir'. Press enter to continue.
```



Either way, I guess we can close this ticket for now. Thanks again!

**Comment by sylviamoss** [comment]
@sfuerte the plugin's name is  `qemu-chroot` as well as the builder's name. For the amazon plugin, the builder's name is `chroot` while the plugin's name is `amazon`, that's why there's no need for the `plugin.DEFAULT_NAME`. 
If we use `plugin.DEFAULT_NAME` when registering the amazon's chroot builder its reference in templates would have to be just `amazon` (the name of the plugin)

**Comment by sylviamoss** [comment]
As for the pause after every step, I believe this is happening because you're using the `-debug` flag.

**Comment by nywilken** [comment]
Given the contents of this thread, and @sfuerte, it seems safe to close this issue. Thanks @sylviamoss for the help in troubleshooting. 

@sfuerte if you find that you are still running into issues please don't hesitate to open a new issue.
